LoopKit/HealthKitSampleStore.swift

```diff
@@ -141,53 +153,108 @@ public class HealthKitSampleStore {
+    }
+    internal let lockedQueryAnchor: Locked<HKQueryAnchor?>
+
+    /// The last observer query completion, stored until the next anchored object query returns
+    private var observerQueryCompletionHandler: HKObserverQueryCompletionHandler? {
+        get {
+            lockedObserverQueryCompletionHandler.value
+        }
+        set {
+            var oldValue: HKObserverQueryCompletionHandler?
+            lockedObserverQueryCompletionHandler.mutate { handler in
+                oldValue = handler
+                handler = newValue
+            }
+
+            oldValue?()
+        }
+    }
+    private let lockedObserverQueryCompletionHandler: Locked<HKObserverQueryCompletionHandler?> = Locked(nil)
+
+    func queryAnchorDidChange() {
+        // Subclasses can override
+    }
 
     /// Called in response to an update by the observer query
     ///
     /// - Parameters:
     ///   - query: The query which triggered the update
     ///   - error: An error during the update, if one occurred
-    internal func observeUpdates(to query: HKObserverQuery, error: Error?) {
-        guard error == nil else {
-            log.error("%@ notified with changes with error: %{public}@", query, String(describing: error))
+    internal final func observeUpdates(to query: HKObserverQuery, completionHandler: HKObserverQueryCompletionHandler?, error: Error?) {
+        if let error = error {
+            self.log.error("%@ notified with changes with error: %{public}@", query, String(describing: error))
+            completionHandler?()
             return
         }
 
-        let anchoredObjectQuery = HKAnchoredObjectQuery(
-            type: self.sampleType,
-            predicate: query.predicate,
-            anchor: self.queryAnchor,
-            limit: HKObjectQueryNoLimit
-        ) { (query, newSamples, deletedSamples, anchor, error) in
-            self.log.debug("%@: new: %d deleted: %d anchor: %@ error: %@", #function, newSamples?.count ?? 0, deletedSamples?.count ?? 0, String(describing: anchor), String(describing: error))
+        // Hold the completion handler (calling any existing ones) until our next anchored object query returns
+        self.observerQueryCompletionHandler = completionHandler
 
-            if let error = error {
-                self.log.error("%@: error executing anchoredObjectQuery: %@", String(describing: type(of: self)), error.localizedDescription)
-            }
+        let queryAnchor = self.queryAnchor
+        log.default("%@ notified with changes. Fetching from: %{public}@", query, queryAnchor.map(String.init(describing:)) ?? "0")
 
-            self.processResults(from: query, added: newSamples ?? [], deleted: deletedSamples ?? [], error: error)
-            self.queryAnchor = anchor
+        let anchoredObjectQuery = createAnchoredObjectQuery(sampleType, queryAnchor == nil ? query.predicate : nil, queryAnchor, HKObjectQueryNoLimit) {  [weak self] (query, newSamples, deletedSamples, anchor, error) in
```

Comment:
Why using a `nil` predicate if a query anchor exists? I don't believe the query anchor captures the previous predicate, but just a pointer into the HealthKit "history".